File: src/go/4_BehavioralPatterns/10_Template/functionaltemplatemethod.go
package templatemethod

import "fmt"

func PlayGame(start, takeTurn func(),
  haveWinner func()bool,
  winningPlayer func()int) {
  start()
  for ;!haveWinner(); {
    takeTurn()
  }
  fmt.Printf("Player %d wins.\n", winningPlayer())
}

func main() {
  turn, maxTurns, currentPlayer := 1, 10, 0

  start := func() {
    fmt.Println("Starting a game of chess.")
  }

  takeTurn := func() {
    turn++
    fmt.Printf("Turn %d taken by player %d\n",
      turn, currentPlayer)
    currentPlayer = (currentPlayer + 1) % 2
  }

  haveWinner := func()bool {
    return turn == maxTurns
  }

  winningPlayer := func()int {
    return currentPlayer
  }

  PlayGame(start, takeTurn, haveWinner, winningPlayer)
}

File: src/csharp/4_BehavioralPatterns/11_Template/FunctionalTemplateMethod.cs
using System;

namespace DotNetDesignPatternDemos.Behavioral.TemplateMethod
{
  using static System.Console;
  
  public static class GameTemplate
  {
    public static void Run(
      Action start,
      Action takeTurn,
      Func<bool> haveWinner,
      Func<int> winningPlayer)
    {
      start();
      while (!haveWinner())
        takeTurn();
      WriteLine($"Player {winningPlayer()} wins.");
    }
  }

  public class Demo2
  {
    static void Main(string[] args)
    {
      var numberOfPlayers = 2;
      int currentPlayer = 0;
      int turn = 1, maxTurns = 10;
      
      void Start()
      {
        WriteLine($"Starting a game of chess with {numberOfPlayers} players.");
      }

      bool HaveWinner()
      {
        return turn == maxTurns;
      }
      
      void TakeTurn()
      {
        WriteLine($"Turn {turn++} taken by player {currentPlayer}.");
        currentPlayer = (currentPlayer + 1) % numberOfPlayers;
      }
      
      int WinningPlayer() {
        return currentPlayer;
      }
      
      GameTemplate.Run(Start, TakeTurn, HaveWinner, WinningPlayer);
    }
  }
}